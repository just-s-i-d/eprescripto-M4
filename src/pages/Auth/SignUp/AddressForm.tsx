import { Form, Input } from "antd";

const AddressForm = () => {
  return (
    <>
      <Form.Item
        label="Country"
        name="country"
        required
        rules={[
          {
            required: true,
            message: "Field cannot be empty",
          },
          {
            pattern: /^[A-Za-z]+$/,
            message: "Number and special characters not allowed",
          },
        ]}
      >
        <Input />
      </Form.Item>
      <Form.Item
        label="State"
        name="state"
        required
        rules={[
          {
            required: true,
            message: "Field cannot be empty",
          },
          {
            pattern: /^[A-Za-z]+$/,
            message: "Number and special characters not allowed",
          },
        ]}
      >
        <Input />
      </Form.Item>
      <Form.Item
        label="City"
        name="city"
        required
        rules={[
          {
            required: true,
            message: "Field cannot be empty",
          },
          {
            pattern: /^[A-Za-z]+$/,
            message: "Number and special characters not allowed",
          },
        ]}
      >
        <Input />
      </Form.Item>
      {/* <Form.Item
          label="Pincode"
          name="pincode"
          required
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
            {
              len: 6,
              message: "Enter a valid pincode",
            },
          ]}
        >
          <Input />
        </Form.Item> */}
    </>
  );
};

export default AddressForm;
