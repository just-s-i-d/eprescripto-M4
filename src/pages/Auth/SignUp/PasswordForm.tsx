import { Form, Input } from "antd";

const PasswordForm = () => {
  return (
    <>
      <Form.Item
        name="password"
        label="Password"
        rules={[
          {
            required: true,
            message: "Field cannot be empty",
          },
          {
            min: 8,
            message: "Password length should be between 8 to 15",
          },
          {
            max: 15,
            message: "Password length should be between 8 to 15",
          },
        ]}
        hasFeedback
      >
        <Input.Password />
      </Form.Item>
      <Form.Item
        name="confirmPassword"
        label="Confirm Password"
        dependencies={["password"]}
        hasFeedback
        rules={[
          {
            required: true,
            message: "Field cannot be empty",
          },
          ({ getFieldValue }) => ({
            validator(_, value) {
              if (!value || getFieldValue("password") === value) {
                return Promise.resolve();
              }
              return Promise.reject(new Error("Password doesn't match!"));
            },
          }),
        ]}
      >
        <Input.Password />
      </Form.Item>
    </>
  );
};
export default PasswordForm;
