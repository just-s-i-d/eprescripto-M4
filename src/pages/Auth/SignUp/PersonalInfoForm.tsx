import { Form, Input, Select } from "antd";

const PersonalInfoForm = () => {
  return (
    <>
      <Form.Item
        label="Name"
        name="docName"
        className="text-xl"
        required
        rules={[
          {
            pattern: /^(Dr\.?|Doctor)?\s?[A-Za-z\s.'-]+$/,
            message: "Special characters not allowed",
          },
          {
            required: true,
            message: "Field cannot be empty",
          },
        ]}
      >
        <Input placeholder="ex. John Doe or Dr. John Doe" />
      </Form.Item>
      <Form.Item
        label="Email"
        name="email"
        required
        rules={[
          {
            required: true,
            message: "Field cannot be empty",
          },
          {
            pattern: /^[\w.-]+@[\w.-]+\.\w+$/,
            message: "Enter a valid mail",
          },
        ]}
      >
        <Input placeholder="ex. John@example.com" />
      </Form.Item>
      <div className="w-full flex justify-between">
        <Form.Item
          label="Gender"
          name="gender"
          required
          className="w-[45%]"
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
          ]}
        >
          <Select
            placeholder="Select a gender"
            options={[
              { value: "male", label: "Male" },
              { value: "female", label: "Female" },
              { value: "other", label: "Other" },
            ]}
          />
        </Form.Item>
        <Form.Item
          label="Date of Birth"
          required
          className="w-[45%]"
          name="dob"
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
          ]}
        >
          <Input type="date" placeholder="ex. 24" />
        </Form.Item>
      </div>
    </>
  );
};

export default PersonalInfoForm;
