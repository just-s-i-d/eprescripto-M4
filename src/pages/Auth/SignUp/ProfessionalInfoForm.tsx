import { Form, Input, Select } from "antd";

const ProfessionalInfoForm = () => {
  return (
    <>
      <Form.Item
        name="licenseNo"
        label="License Number"
        rules={[
          {
            required: true,
            message: "Field cannot be empty",
          },
          {
            len: 4,
            message: "Only 4 digits allowed",
          },
        ]}
      >
        <Input type="number" />
      </Form.Item>
      <div className="w-full flex justify-between">
        <Form.Item
          name="speciality"
          label="Speciality"
          className="w-[45%]"
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
          ]}
        >
          <Input placeholder="ex. Surgon" />
        </Form.Item>
        <Form.Item
          className="w-[45%]"
          name="experience"
          label="Experience"
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
          ]}
        >
          <Input type="number" placeholder="In years" />
        </Form.Item>
      </div>
      <div className="w-full flex justify-between">
        <Form.Item
          name="organizationType"
          label="Organization Type"
          className="w-[45%]"
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
          ]}
        >
          <Select
            placeholder="Select organization type"
            options={[
              { value: "hospital", label: "Hospital" },
              { value: "clinic", label: "Clinic" },
            ]}
          />
        </Form.Item>
        <Form.Item
          name="organizationName"
          label="Organization Name"
          className="w-[45%]"
          rules={[
            {
              required: true,
              message: "Field cannot be empty",
            },
          ]}
        >
          <Input />
        </Form.Item>
      </div>
    </>
  );
};
export default ProfessionalInfoForm;
