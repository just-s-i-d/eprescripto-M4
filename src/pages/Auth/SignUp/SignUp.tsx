import Description from "@components/ui/Description";
import { SIGNUP_FORM_STEPS, STEPPER_STEPS } from "@constants/constants";
import { UserDetailsType } from "@constants/types";
import { Button, Form, Space, Steps } from "antd";
import { SetStateAction, useEffect, useState } from "react";
import CardSider from "../CardSider";
import { checkUserAlreadyExists, registerUser } from "@utils/Doctor";
import { showToast } from "@utils/common";
import useDirectionHook from "@hooks/useDirectionHook";

const { useForm, useWatch } = Form;

type SignUpPropsType = {
  setShowSignIn: React.Dispatch<SetStateAction<boolean>>;
};
const SignUp = ({ setShowSignIn }: SignUpPropsType) => {
  const [current, setCurrent] = useState(0);
  const [userDetails, setUserDetails] = useState<UserDetailsType>();
  const [disabled, setDisabled] = useState(true);
  const [loading, setLoading] = useState(false);
  const [signUpForm] = useForm();
  const formValues = useWatch([], signUpForm);
  const direction = useDirectionHook(787);
  const onFinish = (values: UserDetailsType) => {
    setUserDetails((prev) => ({ ...prev, ...values }));
    if (!current) {
      checkUserAlreadyExists(values.email)
        .then(() => {
          onClickHanlderNext();
        })
        .catch((error) => {
          showToast(error, "error");
        });
    } else {
      onClickHanlderNext();
    }
  };
  const onClickHanlderNext = () => {
    if (current < STEPPER_STEPS.length - 1) setCurrent((prev) => prev + 1);
  };
  const onClickHanlderPrevious = () => {
    if (current > 0) setCurrent((prev) => prev - 1);
  };
  useEffect(() => {
    signUpForm.validateFields({ validateOnly: true }).then(
      () => {
        setDisabled(false);
      },
      () => {
        setDisabled(true);
      },
    );
  }, [formValues]);
  const onSubmitHandler = () => {
    setLoading(true);
    registerUser({ ...userDetails, ...formValues })
      .then((res) => {
        showToast(res.message, res.type);
      })
      .catch((error) => {
        showToast(error.message, error.type);
      })
      .finally(() => {
        setLoading(false);
      });
  };
  return (
    <>
      <CardSider hidden={true}>
        <Steps
          className="xs:mt-8 md:mt-0 md:h-[35vh]"
          items={STEPPER_STEPS}
          current={current}
          direction={direction}
          responsive={false}
        />
      </CardSider>
      <Space direction="vertical" className="w-full h-5/6 px-8">
        <div className="text-[34px]">Create Account</div>
        <div>
          <h2 className="text-xl font-bold mb-1">
            {SIGNUP_FORM_STEPS[current].title}
          </h2>
          <Description>{SIGNUP_FORM_STEPS[current].description}</Description>
        </div>
        <Form
          form={signUpForm}
          className="custom-form flex flex-col gap-4"
          size="large"
          layout="vertical"
          onFinish={onFinish}
          data-testid="sign-up-form"
          initialValues={userDetails}
        >
          {SIGNUP_FORM_STEPS[current].content}
          <Space className="absolute bottom-8 right-8">
            {current > 0 && (
              <Button
                className="text-white w-[90px] bg-primary"
                onClick={onClickHanlderPrevious}
              >
                Previous
              </Button>
            )}
            {current < STEPPER_STEPS.length - 1 && (
              <Button
                className="text-white w-[90px] bg-primary disabled:opacity-50"
                htmlType="submit"
                disabled={disabled}
                loading={loading}
              >
                Next
              </Button>
            )}
            {current === STEPPER_STEPS.length - 1 && (
              <Button
                className="text-white w-[90px] bg-primary disabled:opacity-50"
                onClick={onSubmitHandler}
                disabled={disabled}
                data-testid="submit-btn"
              >
                Submit
              </Button>
            )}
          </Space>
        </Form>
        {!current && (
          <div className="absolute bottom-[36px]">
            Already have an account ?
            <span
              className="text-base cursor-pointer text-blue-500 ml-2"
              onClick={() => setShowSignIn(true)}
            >
              Sign In
            </span>
          </div>
        )}
      </Space>
    </>
  );
};
export default SignUp;
