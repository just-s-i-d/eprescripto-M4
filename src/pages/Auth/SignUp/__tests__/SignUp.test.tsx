import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import SignUp from "../SignUp";
import "@testing-library/jest-dom";
import "@test/matchMedia";
import { act } from "react-dom/test-utils";

jest.mock("@utils/Doctor", () => ({
  ...jest.requireActual("@utils/Doctor"),
  checkUserAlreadyExists: jest.fn().mockResolvedValue("Mocked"),
}));
describe("Sign up tests", () => {
  const setShowSignInMocked = jest.fn();
  it("Render the button to switch to sign in form", () => {
    const { asFragment } = render(
      <SignUp setShowSignIn={setShowSignInMocked} />,
    );
    const signInBtn = screen.getByText("Sign In");
    expect(signInBtn).toBeInTheDocument();
    act(() => {
      fireEvent.click(signInBtn);
    });
    expect(setShowSignInMocked).toHaveBeenCalled();
    expect(asFragment()).toMatchSnapshot();
  });

  it("Test to check User email already exists", async () => {
    render(<SignUp setShowSignIn={setShowSignInMocked} />);
    const genderSelect = screen.getByRole("combobox", { name: "Gender" });
    expect(genderSelect).toBeInTheDocument();
    fireEvent.change(genderSelect, { target: { value: "male" } });
    await waitFor(() => {
      const optionMale = document.getElementById("gender_list_0");
      optionMale && fireEvent.click(optionMale);
      expect(optionMale).toBeInTheDocument();
    });
    await waitFor(() => {
      const nextBtn = screen.getByRole("button", { name: "Next" });
      expect(nextBtn).toBeDisabled();
    });
  });

  it("User registration test", async () => {
    render(<SignUp setShowSignIn={setShowSignInMocked} />);
    const name = screen.getByLabelText("Name");
    const email = screen.getByLabelText("Email");
    const gender = screen.getByLabelText("Gender");
    const dob = screen.getByLabelText("Date of Birth");
    act(() => {
      fireEvent.change(name, { target: { value: "John" } });
      fireEvent.change(email, { target: { value: "email@email.com" } });
      fireEvent.select(gender, { target: { value: "male" } });
      fireEvent.change(dob, { target: { value: "2001-05-27" } });
    });
    expect(name).toHaveValue("John");
    expect(email).toHaveValue("email@email.com");
    expect(gender).toHaveValue("male");
    expect(dob).toHaveValue("2001-05-27");
    const nextBtn = screen.getByRole("button", { name: "Next" });
    fireEvent.click(nextBtn);
  });
});
