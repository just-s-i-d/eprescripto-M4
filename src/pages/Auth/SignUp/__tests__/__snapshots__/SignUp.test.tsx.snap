// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign up tests Render the button to switch to sign in form 1`] = `
<DocumentFragment>
  <div
    class="ant-space css-dev-only-do-not-override-1adbn6x ant-space-vertical ant-space-gap-row-small ant-space-gap-col-small xs:w-full xs:mb-8 xs:h-[20%] md:mb-0 md:w-[45%] bg-primary rounded-[30px] px-6 md:h-full flex justify-center items-center"
  >
    <div
      class="ant-space-item"
    >
      <h1
        class="text-center text-4xl text-white mb-4 xs:mt-4 md:mt-0 xs:hidden md:block"
      >
        ePrescripto
      </h1>
    </div>
    <div
      class="ant-space-item"
    >
      <div
        class="ant-steps ant-steps-vertical xs:mt-8 md:mt-0 md:h-[35vh] css-dev-only-do-not-override-1adbn6x"
      >
        <div
          class="ant-steps-item ant-steps-item-process ant-steps-item-active"
        >
          <div
            class="ant-steps-item-container"
          >
            <div
              class="ant-steps-item-tail"
            />
            <div
              class="ant-steps-item-icon"
            >
              <span
                class="ant-steps-icon"
              >
                1
              </span>
            </div>
            <div
              class="ant-steps-item-content"
            >
              <div
                class="ant-steps-item-title"
              >
                Step 1
              </div>
              <div
                class="ant-steps-item-description"
              >
                Personal Info
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-steps-item ant-steps-item-wait"
        >
          <div
            class="ant-steps-item-container"
          >
            <div
              class="ant-steps-item-tail"
            />
            <div
              class="ant-steps-item-icon"
            >
              <span
                class="ant-steps-icon"
              >
                2
              </span>
            </div>
            <div
              class="ant-steps-item-content"
            >
              <div
                class="ant-steps-item-title"
              >
                Step 2
              </div>
              <div
                class="ant-steps-item-description"
              >
                Personal Address
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-steps-item ant-steps-item-wait"
        >
          <div
            class="ant-steps-item-container"
          >
            <div
              class="ant-steps-item-tail"
            />
            <div
              class="ant-steps-item-icon"
            >
              <span
                class="ant-steps-icon"
              >
                3
              </span>
            </div>
            <div
              class="ant-steps-item-content"
            >
              <div
                class="ant-steps-item-title"
              >
                Step 3
              </div>
              <div
                class="ant-steps-item-description"
              >
                Professional Info
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-steps-item ant-steps-item-wait"
        >
          <div
            class="ant-steps-item-container"
          >
            <div
              class="ant-steps-item-tail"
            />
            <div
              class="ant-steps-item-icon"
            >
              <span
                class="ant-steps-icon"
              >
                4
              </span>
            </div>
            <div
              class="ant-steps-item-content"
            >
              <div
                class="ant-steps-item-title"
              >
                Step 4
              </div>
              <div
                class="ant-steps-item-description"
              >
                Finish
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="ant-space css-dev-only-do-not-override-1adbn6x ant-space-vertical ant-space-gap-row-small ant-space-gap-col-small w-full h-5/6 px-8"
  >
    <div
      class="ant-space-item"
    >
      <div
        class="text-[34px]"
      >
        Create Account
      </div>
    </div>
    <div
      class="ant-space-item"
    >
      <div>
        <h2
          class="text-xl font-bold mb-1"
        >
          Personal Information
        </h2>
        <h1
          class="text-sm font-bold mb-1 text-slate-400 undefined"
        >
          Please enter your personal information
        </h1>
      </div>
    </div>
    <div
      class="ant-space-item"
    >
      <form
        class="ant-form ant-form-vertical ant-form-large css-dev-only-do-not-override-1adbn6x custom-form flex flex-col gap-4"
        data-testid="sign-up-form"
      >
        <div
          class="ant-form-item text-xl css-dev-only-do-not-override-1adbn6x"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-1adbn6x"
          >
            <div
              class="ant-col ant-form-item-label css-dev-only-do-not-override-1adbn6x"
            >
              <label
                class="ant-form-item-required"
                for="docName"
                title="Name"
              >
                Name
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-1adbn6x"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    aria-required="true"
                    class="ant-input ant-input-lg css-dev-only-do-not-override-1adbn6x"
                    id="docName"
                    placeholder="ex. John Doe or Dr. John Doe"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-form-item css-dev-only-do-not-override-1adbn6x"
        >
          <div
            class="ant-row ant-form-item-row css-dev-only-do-not-override-1adbn6x"
          >
            <div
              class="ant-col ant-form-item-label css-dev-only-do-not-override-1adbn6x"
            >
              <label
                class="ant-form-item-required"
                for="email"
                title="Email"
              >
                Email
              </label>
            </div>
            <div
              class="ant-col ant-form-item-control css-dev-only-do-not-override-1adbn6x"
            >
              <div
                class="ant-form-item-control-input"
              >
                <div
                  class="ant-form-item-control-input-content"
                >
                  <input
                    aria-required="true"
                    class="ant-input ant-input-lg css-dev-only-do-not-override-1adbn6x"
                    id="email"
                    placeholder="ex. John@example.com"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="w-full flex justify-between"
        >
          <div
            class="ant-form-item w-[45%] css-dev-only-do-not-override-1adbn6x"
          >
            <div
              class="ant-row ant-form-item-row css-dev-only-do-not-override-1adbn6x"
            >
              <div
                class="ant-col ant-form-item-label css-dev-only-do-not-override-1adbn6x"
              >
                <label
                  class="ant-form-item-required"
                  for="gender"
                  title="Gender"
                >
                  Gender
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control css-dev-only-do-not-override-1adbn6x"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <div
                      aria-required="true"
                      class="ant-select ant-select-lg ant-select-in-form-item css-dev-only-do-not-override-1adbn6x ant-select-single ant-select-show-arrow"
                      data-testid="gender-select-signup"
                    >
                      <div
                        class="ant-select-selector"
                      >
                        <span
                          class="ant-select-selection-search"
                        >
                          <input
                            aria-autocomplete="list"
                            aria-controls="gender_list"
                            aria-expanded="false"
                            aria-haspopup="listbox"
                            aria-owns="gender_list"
                            aria-required="true"
                            autocomplete="off"
                            class="ant-select-selection-search-input"
                            id="gender"
                            readonly=""
                            role="combobox"
                            style="opacity: 0;"
                            type="search"
                            unselectable="on"
                            value=""
                          />
                        </span>
                        <span
                          class="ant-select-selection-placeholder"
                        >
                          Select a gender
                        </span>
                      </div>
                      <span
                        aria-hidden="true"
                        class="ant-select-arrow"
                        style="user-select: none;"
                        unselectable="on"
                      >
                        <span
                          aria-label="down"
                          class="anticon anticon-down ant-select-suffix"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="down"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                            />
                          </svg>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-form-item w-[45%] css-dev-only-do-not-override-1adbn6x"
          >
            <div
              class="ant-row ant-form-item-row css-dev-only-do-not-override-1adbn6x"
            >
              <div
                class="ant-col ant-form-item-label css-dev-only-do-not-override-1adbn6x"
              >
                <label
                  class="ant-form-item-required"
                  for="dob"
                  title="Date of Birth"
                >
                  Date of Birth
                </label>
              </div>
              <div
                class="ant-col ant-form-item-control css-dev-only-do-not-override-1adbn6x"
              >
                <div
                  class="ant-form-item-control-input"
                >
                  <div
                    class="ant-form-item-control-input-content"
                  >
                    <input
                      aria-required="true"
                      class="ant-input ant-input-lg css-dev-only-do-not-override-1adbn6x"
                      data-testid="date-signup"
                      id="dob"
                      placeholder="ex. 24"
                      type="date"
                      value=""
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-space css-dev-only-do-not-override-1adbn6x ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small absolute bottom-8 right-8"
        >
          <div
            class="ant-space-item"
          >
            <button
              class="ant-btn css-dev-only-do-not-override-1adbn6x ant-btn-default ant-btn-lg text-white w-[90px] bg-primary disabled:opacity-50"
              disabled=""
              type="submit"
            >
              <span>
                Next
              </span>
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      class="ant-space-item"
    >
      <div
        class="absolute bottom-[36px]"
      >
        Already have an account ?
        <span
          class="text-base cursor-pointer text-blue-500 ml-2"
        >
          Sign In
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
